Add documentation for every state selector

aduth opened the pull request:
This pull request seeks to add documentation for every state selector which exists. Some are a best guess at intent of the selector, so review from original authors would be appreciated.

__Testing instructions:__

There is no impact on editor experience from these changes.

Ensure that new unit tests pass:

```
npm run test-unit
```

aduth committed:
Document every selector

nylen commented on editor/test/selectors.js:
```diff
@@ -839,4 +890,57 @@ describe( 'selectors', () => {
 	} );
+
+	describe( 'getSuggestedPostFormat', () => {
+		it( 'returns null if cannot be determined', () => {
```
Wording of these test names? Most of our other tests use `should`, and it seems good to be consistent... 

```js
it( 'should be painted yellow', ...
```

aduth replied:
>Wording of these test names?

We just talked about this at https://github.com/WordPress/gutenberg/pull/946#issuecomment-305277136 😛  I'd argue this is one case where I wish we _weren't_ so consistent, because bending over backwards to prefix with "should" isn't always appropriate and at worst is unnecessary filler.

nylen commented on editor/test/selectors.js:
```diff
@@ -459,6 +462,54 @@ describe( 'selectors', () => {
+				editor: {
+					blockOrder: [ 5, 4, 3, 2, 1 ],
+				},
+				multiSelectedBlocks: { start: 2, end: 4 },
```
Here and below, it seems strange to have `enf` before `start` in the block order. Is this possible in actual usage?

iseulde replied:
Yes, `start` has the index where the selection started, `end` where it ended. See also `getSelectedBlocks`.

iseulde replied:
(You can select blocks from bottom to top.)

nylen commented on editor/selectors.js:
```diff
@@ -8,48 +8,121 @@ import { first, last, get } from 'lodash';
 }
 
+/**
+ * Returns true if the past editor history snapshots exist, or false otherwise.
```
s/the/any/ - it reads better to me this way

nylen commented on editor/selectors.js:
```diff
@@ -8,48 +8,121 @@ import { first, last, get } from 'lodash';
 }
 
+/**
+ * Returns true if the future editor history snapshots exist, or false
```
s/the/any/ - it reads better to me this way

nylen commented on editor/selectors.js:
```diff
@@ -8,48 +8,121 @@ import { first, last, get } from 'lodash';
 }
 
+/**
+ * Returns the post currently being edited.
```
Probably worth noting that this does not include any edits made to the post. "Returns our representation of the post in its last known state state on the server" maybe?

aduth replied:
>Probably worth noting that this does not include any edits made to the post. "Returns our representation of the post in its last known state state on the server" maybe?

Raises an interesting point about what we'd expect the return value to be for a new post. Based on the [`currentPost` default](https://github.com/WordPress/gutenberg/blob/1e464894e30afae275895a1c40d4c124da3f30fd/editor/state.js#L184) I'd guess an empty object? Seems like it might be more appropriate as `null`, though I suspect this could cause breakage where we expect to access properties unguarded on the post object.

Nice thing about adding these documentation is that it helps surface real-world expectations 😄

nylen replied:
Not necessarily an empty object - I imagine a new post would contain `status: 'draft'` and any other default values.  Previous discussion: https://github.com/WordPress/gutenberg/pull/848#issuecomment-302834261

Currently these default values are stored in `post-content.js` ([#](https://github.com/WordPress/gutenberg/blob/728c17761585efbbd31ebef870b64428a7c88861/post-content.js#L8), [#](https://github.com/WordPress/gutenberg/blob/728c17761585efbbd31ebef870b64428a7c88861/editor/index.js#L30)).

So perhaps we could improve the wording here as follows:

> Returns our representation of the saved post object.  For an existing post, this is the last known state of the post on the server; for a new post, this is an object containing any relevant default values.

nylen commented on editor/selectors.js:
```diff
@@ -8,48 +8,121 @@ import { first, last, get } from 'lodash';
 }
 
+/**
+ * Returns the post values which have yet to be saved.
```
"Returns any post values that have been changed in the editor and not yet saved"?

nylen commented on editor/selectors.js:
```diff
@@ -8,48 +8,121 @@ import { first, last, get } from 'lodash';
 }
 
+/**
+ * Returns a single attribute of the post being edited.
```
"..., pulling from either the last known representation of the post or its edited value, if any."

nylen commented on editor/selectors.js:
```diff
@@ -122,44 +245,124 @@ export function getSelectedBlocks( state ) {
+ * @param  {String} uid   Block unique ID
+ * @return {Number}       Index at which block exists in order
+ */
 export function getBlockOrder( state, uid ) {
```
Perhaps separately, I think this selector should be renamed. As-is, I'd expect it to simply​ return the block order array. `getBlockIndex` maybe?

nylen commented on editor/selectors.js:
```diff
@@ -170,14 +373,39 @@ export function isBlockSelected( state, uid ) {
+/**
+ * Returns focus state of the block corresponding to the specified unique ID,
+ * or null if the block is not selected. Focus state defaults to an empty
+ * object.
```
What other object shapes might be returned by the focus state?

aduth replied:
>What other object shapes might be returned by the focus state?

As far as I recall the original intention, it's up to the block to manage the focus object how they see fit. For example differentiating between Editables of the same block:

https://github.com/WordPress/gutenberg/blob/1e46489/blocks/library/quote/index.js#L115

This is one thing I'd like to see refactored, ideally with less block implementer intervention. Relates to #878

cc @youknowriad

nylen replied:
Agree this could be improved, for now I would add something like "The specific content of the object is left up to the implementation of each block."

nylen commented on editor/selectors.js:
```diff
@@ -194,6 +430,14 @@ export function isTypingInBlock( state, uid ) {
 }
 
+/**
+ * Returns the unique ID of the block at which a new block insertion would be
```
s/at which/after which/ ?

nylen reviewed (commented):
Thanks for doing this. I learned a few things about our state structure already from reviewing it.

youknowriad commented on editor/selectors.js:
```diff
@@ -8,48 +8,121 @@ import { first, last, get } from 'lodash';
+ * @param  {Object} state Global application state
+ * @return {String}       Post status
+ */
 export function getEditedPostStatus( state ) {
```
This has been written before the `getEditedPostAttribute`, I wonder if we should just drop it now.

youknowriad commented on editor/selectors.js:
```diff
@@ -122,44 +245,124 @@ export function getSelectedBlocks( state ) {
 }
 
+/**
+ * Returns an array containing all unique IDs of the post being edited, in the
```
all the unique block IDS

aduth committed:
Rename getBlockOrder as getBlockIndex

aduth committed:
Remove getEditedPostStatus selector

In favor of direct `getEditedPostAttribute( state, 'status' )`

aduth commented:
Revisions made per reviews. Thanks for looking 👍

youknowriad reviewed (approved):
Thanks for doing this

aduth commented:
Updated per feedback at https://github.com/WordPress/gutenberg/pull/989#discussion_r119902454 and https://github.com/WordPress/gutenberg/pull/989#discussion_r119903001, squashed the two revisory rewording commits. Will merge when CI passes.
